# Tutorial 3: A First Model

After getting a basic understanding of the data in the previous tutorial, we are ready to build a first AI model.
In this notebook, we build and evaluate a simple model for identifying worm sections in nodule images.

By the end of this notebook you will have created a first submission that you can submit to put your team on the leaderboard!

**NOTE: This notebook will only work with a GPU instance. Make sure that you selected *ml.g4dn.xlarge* as instance type**

# Picking the right approach

The first challenge we have to address is how to even tackle the problem. What could be a good approach / algorithm? How are other people doing it?
[PapersWithCode](https://paperswithcode.com/sota) is an amazing resource for exactly these questions. It lists typical problem classes together with the current state-of-the-art and links to popular frameworks.
For our problem, [Object Detection](https://paperswithcode.com/task/object-detection) is the right category. If you follow the link you'll find a list of popular frameworks.

[MMDetection](https://github.com/open-mmlab/mmdetection) is one of the most widely used and the one we picked for this tutorial. Not only does it provide a great [tutorial](https://github.com/open-mmlab/mmdetection/blob/master/demo/MMDet_Tutorial.ipynb), but also offers a variety of [different architectures and settings](https://github.com/open-mmlab/mmdetection#overview-of-benchmark-and-model-zoo) that you can play around with.

In this notebook, we will adapt the MMdetection [tutorial](https://github.com/open-mmlab/mmdetection/blob/master/demo/MMDet_Tutorial.ipynb) to our problem and run it on a toy dataset.

For those familiar with deep learning architectures, MMDetection can handle data loading, preperation, model building, and training by altering a few common settings in the configuration file. Data preperation also includes various data augmentation techinques like rotation or flipping. The majority of the most common CV arcitectures are already build and available via configuration.

### *Further reading:*
Going into details of object detection is beyond the scope of this tutorial. If you'd like to know more, here are two recommendations to get started:
- [An introduction to R-CNNs and object detection in general](https://towardsdatascience.com/deep-dive-into-the-computer-vision-world-part-2-7a24efdb1a14)
- [An graphic explanation how convolutional neural networks work](https://towardsdatascience.com/gentle-dive-into-math-behind-convolutional-neural-networks-79a07dd44cf9)

# Setup
As in the other notebooks, we start by importing the relevant libraries and global settings.

In [1]:
import matplotlib.pyplot as plt  # Used for plotting
import mmcv  # Object detection framework
import os  # Interaction with the file system
import pandas as pd  # Home of the DataFrame construct, _the_ most important object for Data Science
import sys  # Python system library needed to load custom functions

from matplotlib.patches import Rectangle  # Allows drawing the bounding boxes of the worm sections
from mmcv import Config  # Loading and accessing MMDetection configuration files
from mmdet.apis import inference_detector, init_detector, train_detector, set_random_seed  # Part of the MMDetection framework
from mmdet.datasets import build_dataset  # Part of the MMDetection framework
from mmdet.models import build_detector  # Part of the MMDetection framework

from PIL import Image  # For loading image files
from tqdm import tqdm  # for timing a for loop

In [2]:
sys.path.append('../src')  # Add the source directory to the PYTHONPATH. This allows to import local functions and modules.

In [3]:
from Dataset import OnchoDataset
from detection_util import create_predictions
from gdsc_score import get_leaderboard_score
from gdsc_util import download_directory, download_file, load_sections_df, set_up_logging, PROJECT_DIR
from PredictionEvaluator import PredictionEvaluator

set_up_logging()  # Sets up logging to console and the .log file
data_folder = str(PROJECT_DIR / 'data')

# Preparing the Training Dataset

Whenever you start working with a new framework you can expect issues with the setup.
To ensure that we can quickly try things out we create a small subset of our data.

In [4]:
section_df = load_sections_df(f'{data_folder}/gdsc_train.csv')
len(section_df)

65687

Our complete dataset has more than 65.000 rows. Let's create two small dummy sets with 100 and 50 entries only.

In [5]:
dummy_train = section_df[:100]
dummy_test = section_df[100:150]
dummy_train.to_csv(f'{data_folder}/dummy_train.csv', sep=';')
dummy_test.to_csv(f'{data_folder}/dummy_test.csv', sep=';')

We store the reduced sets under *../data*. Verify that you can find and open the files there. 
To make sure that everything works, let's load the files again.

In [6]:
dummy_train = load_sections_df(f'{data_folder}/dummy_train.csv')
dummy_train.head()

file_name    study staining  xmin  xmax  ymin  ymax  \
section_id                                                                    
1_D@2317-5556-2836-6232   1_D.jpg  Study_1        D  2317  2836  5556  6232   
1_D@2407-6156-2952-6789   1_D.jpg  Study_1        D  2407  2952  6156  6789   
1_D@2483-5836-4279-7287   1_D.jpg  Study_1        D  2483  4279  5836  7287   
1_D@2546-1211-4374-2558   1_D.jpg  Study_1        D  2546  4374  1211  2558   
1_D@2695-5530-3238-6179   1_D.jpg  Study_1        D  2695  3238  5530  6179   

                         height  width  
section_id                              
1_D@2317-5556-2836-6232    8192   7380  
1_D@2407-6156-2952-6789    8192   7380  
1_D@2483-5836-4279-7287    8192   7380  
1_D@2546-1211-4374-2558    8192   7380  
1_D@2695-5530-3238-6179    8192   7380

Looks good! 

# Creating a MMDetection Configuration File

MMDetection relies on extensive configuration files. The usual process is to adapt an already existing configuration file and apply transfer learning.
To do this, we first need to download the configuration and weights files.

In [7]:
# Make a new folder in the data folder
!mkdir ../data/checkpoints 
# Install wget, a program for downloading files from the internet
!apt update
!apt install wget
# Download the config and weights file
!wget -c https://download.openmmlab.com/mmdetection/v2.0/mask_rcnn/mask_rcnn_r50_caffe_fpn_mstrain-poly_3x_coco/mask_rcnn_r50_caffe_fpn_mstrain-poly_3x_coco_bbox_mAP-0.408__segm_mAP-0.37_20200504_163245-42aa3d00.pth -O ../data/checkpoints/mask_rcnn_r50_caffe_fpn_mstrain-poly_3x_coco_bbox_mAP-0.408__segm_mAP-0.37_20200504_163245-42aa3d00.pth

mkdir: cannot create directory ‘../data/checkpoints’: File exists
Hit:1 http://archive.ubuntu.com/ubuntu bionic InRelease
Hit:2 http://archive.ubuntu.com/ubuntu bionic-updates InRelease                
Hit:3 http://archive.ubuntu.com/ubuntu bionic-backports InRelease              
Get:4 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease [1581 B]
Hit:5 http://security.ubuntu.com/ubuntu bionic-security InRelease              
Ign:6 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease
Hit:7 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release
Err:4 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY A4B469963BF863CC
Reading package lists... Done0m0m
W: GPG error: https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64 

The configuration file is part of the mmdetection package that is already installed on this image. Hence we can load our test configuration via

In [8]:
cfg = Config.fromfile('/mmdetection/configs/faster_rcnn/faster_rcnn_r50_caffe_fpn_mstrain_1x_coco.py')

The configuration is a dictionary-like object that stores our configuration settings. Is has [a lot](https://mmdetection.readthedocs.io/en/latest/tutorials/config.html) of settings and with the right options, you can run the [really advanced models](https://github.com/open-mmlab/mmdetection/blob/master/docs/en/model_zoo.md).
Below we modify some of those settings to point the model to our train and test data.

We start by defining the structure and location of our data.

In [9]:
# Modify dataset type and path
cfg.dataset_type = 'OnchoDataset' #this is a custom data loader script we created for the GDSC you can view it in src/Dataset.py
cfg.data_root = data_folder

cfg.data.train.type = 'OnchoDataset'
cfg.data.train.data_root = data_folder      # path to the folder data
cfg.data.train.img_prefix = 'jpgs/'         # path from data_root to the images folder
cfg.data.train.ann_file = 'dummy_train.csv' # the file containing the train data labels

cfg.data.test.type = 'OnchoDataset'
cfg.data.test.data_root = data_folder
cfg.data.test.img_prefix = 'jpgs/'
cfg.data.test.ann_file = 'dummy_test.csv'

cfg.data.val.type = 'OnchoDataset'          # We will not use a separate validation data set in this tutorial, but we need to specify the values to overwrite the COCO defaults.
cfg.data.val.data_root = data_folder
cfg.data.val.img_prefix = 'jpgs/'
cfg.data.val.ann_file = 'dummy_test.csv'

We specify where to save the results and where the weights we will use were downloaded to. We also set the number of workers.

In [10]:
# We can still use the pre-trained Mask RCNN model though we do not need to use the mask branch
cfg.load_from = f'{data_folder}/checkpoints/mask_rcnn_r50_caffe_fpn_mstrain-poly_3x_coco_bbox_mAP-0.408__segm_mAP-0.37_20200504_163245-42aa3d00.pth'

# Set up working dir to save files and logs.
cfg.work_dir = f'{data_folder}/tutorial_exps/'
# Ensure work_dir exists
mmcv.mkdir_or_exist(os.path.abspath(cfg.work_dir))

# Set seed thus the results are more reproducible
cfg.seed = 0
set_random_seed(0, deterministic=False)
cfg.gpu_ids = range(1)

cfg.data.samples_per_gpu = 3 # These numbers will change depending on the size of your model and GPU.
cfg.data.workers_per_gpu = 1 # These values are what we have found to be best for this model and GPU

cfg.device = 'cuda'

Next are the model settings. One thing we have to change is the number of classes to predict. Since we only have one class (worm section), we need to set this to one.
We also change the learning rate since we only have one GPU.

In [11]:
# modify number of classes of the model in box head
cfg.model.roi_head.bbox_head.num_classes = 1  # a worm section is the only object we are detecting

# The original learning rate (LR) is set for 8-GPU training.
# We divide it by 8 since we only use one GPU and multiply by the number of GPU workers.
cfg.optimizer.lr = 0.02 / 8 * cfg.data.workers_per_gpu
cfg.lr_config.warmup = None
cfg.log_config.interval = 10

Finally, we specify how and when to evaluate the model and how long to train the model. We will use the [mAP](https://blog.paperspace.com/mean-average-precision/#:~:text=To%20evaluate%20object%20detection%20models,model%20is%20in%20its%20detections.) metric to evaluate how the training is going.

For our test run, we will only train for three epochs. For a real training run you'll want to set this number higher.
After each epoch, the current weights will be saved in the *work_dir* folder we set above. Also the model will be evaluated on the test data set.
This allows us to check if we should continue training or stop.

In [12]:
# Change the evaluation metric since we use customized dataset.
cfg.evaluation.metric = 'mAP'
# We can set the evaluation interval to reduce the evaluation times
cfg.evaluation.interval = 1
# We can set the checkpoint saving interval to reduce the storage cost
cfg.checkpoint_config.interval = 1
# How long do we want to train
cfg.runner.max_epochs = 3

# Training of the first Model

We're now ready to train a first simple model! The first step is to prepare the dataset.

In [13]:
datasets = [build_dataset(cfg.data.train)]

datasets

2022-05-30 14:10:36,785 - Dataset - INFO - Building Dataset
/mmdetection/mmdet/datasets/custom.py:180: UserWarning: CustomDataset does not support filtering empty gt images.
  'CustomDataset does not support filtering empty gt images.')


[
 OnchoDataset Train dataset with number of images 3, and instance counts: 
 +-------------+-------+----------+-------+----------+-------+----------+-------+----------+-------+
 | category    | count | category | count | category | count | category | count | category | count |
 +-------------+-------+----------+-------+----------+-------+----------+-------+----------+-------+
 |             |       |          |       |          |       |          |       |          |       |
 | 0 [section] | 100   |          |       |          |       |          |       |          |       |
 +-------------+-------+----------+-------+----------+-------+----------+-------+----------+-------+]

The dataset contains exactly 100 worm sections, one for each row on the *dummy_train* DataFrame. We also get a warning about filtering empty ground truth images which we can ignore since we don't have those.

Next, we initialize the model according to the configuration. *train_cfg* determins on which data to train the model, *test_cfg* on which data to test the model.

In [14]:
model = build_detector(cfg.model, train_cfg=cfg.get('train_cfg'), test_cfg=cfg.get('test_cfg'))
model.CLASSES = datasets[0].CLASSES  # Add an attribute for visualization convenience

And finally, we can train the model. The log output helps us understand how well the model is doing. 
Since we set the *evaluation.interval* to *1*, the model will be evaluated after every epoch. In general you want to continue training as long as the *mAP* score increases.

In [15]:
train_detector(model, datasets, cfg, validate=True)

2022-05-30 14:10:39,192 - Dataset - INFO - Building Dataset
2022-05-30 14:10:39,205 - mmdet - INFO - load checkpoint from local path: /root/data/AmazonSageMaker-gdsc-tutorials/data/checkpoints/mask_rcnn_r50_caffe_fpn_mstrain-poly_3x_coco_bbox_mAP-0.408__segm_mAP-0.37_20200504_163245-42aa3d00.pth
2022-05-30 14:10:39,707 - mmdet - WARNING - The model and loaded state dict do not match exactly

size mismatch for roi_head.bbox_head.fc_cls.weight: copying a param with shape torch.Size([81, 1024]) from checkpoint, the shape in current model is torch.Size([2, 1024]).
size mismatch for roi_head.bbox_head.fc_cls.bias: copying a param with shape torch.Size([81]) from checkpoint, the shape in current model is torch.Size([2]).
size mismatch for roi_head.bbox_head.fc_reg.weight: copying a param with shape torch.Size([320, 1024]) from checkpoint, the shape in current model is torch.Size([4, 1024]).
size mismatch for roi_head.bbox_head.fc_reg.bias: copying a param with shape torch.Size([320]) from ch

[2022-05-30 14:10:41.937 gdsc5-smstudio-cust-ml-g4dn-xlarge-8e4f662689f9518ad1d6dbca9f90:1295 INFO utils.py:27] RULE_JOB_STOP_SIGNAL_FILENAME: None
[2022-05-30 14:10:41.973 gdsc5-smstudio-cust-ml-g4dn-xlarge-8e4f662689f9518ad1d6dbca9f90:1295 INFO profiler_config_parser.py:102] Unable to find config at /opt/ml/input/config/profilerconfig.json. Profiler is disabled.


2022-05-30 14:10:48,008 - mmdet - INFO - Saving checkpoint at 1 epochs


[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 3/3, 0.9 task/s, elapsed: 3s, ETA:     0s
---------------iou_thr: 0.5---------------


2022-05-30 14:10:55,763 - mmdet - INFO - 
+---------+-----+------+--------+-------+
| class   | gts | dets | recall | ap    |
+---------+-----+------+--------+-------+
| section | 50  | 300  | 0.580  | 0.132 |
+---------+-----+------+--------+-------+
| mAP     |     |      |        | 0.132 |
+---------+-----+------+--------+-------+
2022-05-30 14:10:55,765 - mmdet - INFO - Epoch(val) [1][3]	AP50: 0.1320, mAP: 0.1320
2022-05-30 14:11:01,690 - mmdet - INFO - Saving checkpoint at 2 epochs


[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 3/3, 0.9 task/s, elapsed: 3s, ETA:     0s
---------------iou_thr: 0.5---------------


2022-05-30 14:11:09,386 - mmdet - INFO - 
+---------+-----+------+--------+-------+
| class   | gts | dets | recall | ap    |
+---------+-----+------+--------+-------+
| section | 50  | 300  | 0.760  | 0.250 |
+---------+-----+------+--------+-------+
| mAP     |     |      |        | 0.250 |
+---------+-----+------+--------+-------+
2022-05-30 14:11:09,388 - mmdet - INFO - Epoch(val) [2][3]	AP50: 0.2500, mAP: 0.2502
2022-05-30 14:11:15,307 - mmdet - INFO - Saving checkpoint at 3 epochs


[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 3/3, 0.9 task/s, elapsed: 3s, ETA:     0s
---------------iou_thr: 0.5---------------


2022-05-30 14:11:23,281 - mmdet - INFO - 
+---------+-----+------+--------+-------+
| class   | gts | dets | recall | ap    |
+---------+-----+------+--------+-------+
| section | 50  | 300  | 0.820  | 0.373 |
+---------+-----+------+--------+-------+
| mAP     |     |      |        | 0.373 |
+---------+-----+------+--------+-------+
2022-05-30 14:11:23,285 - mmdet - INFO - Epoch(val) [3][3]	AP50: 0.3730, mAP: 0.3725


The log starts with a warning about a size mismatch that we can ignore. It is due to the fact that the weights we load belong to a model that was trained on the COCO dataset which has 80 different classes whereas we only have one class to detect.

Since the *mAP* keep increasing it would make sense to continue training the model for longer. 
But before we do that let's make sure that the rest of the process works.

**Note**: 
- **If you get an error, try rerunning the notebook starting with the loading of the config.** 
- **If you get the error *RuntimeError: Expected 4-dimensional input for 4-dimensional weight [64, 3, 7, 7], but got 5-dimensional input of size [1, 3, 3, 928, 768] instead* change the instance to one with a GPU.** 



# Running the Model

After training the model we can now apply it on a new image and look at the predictions. 
For this, we first load the image and initalize the model from a stored checkpoint.

In [16]:
example_image = '1_D.jpg'
img =  mmcv.imread(f'{data_folder}/jpgs/{example_image}')

In [17]:
checkpoint = f'{cfg.work_dir}epoch_3.pth' # Select one of the model checkpoints to load in
model = init_detector(cfg, checkpoint, device='cuda')

load checkpoint from local path: /root/data/AmazonSageMaker-gdsc-tutorials/data/tutorial_exps/epoch_3.pth


Run the inference

In [18]:
detections = inference_detector(model, img)
print(len(detections))  
detections[0][:2]  # Show the first two entries

1


/mmdetection/mmdet/datasets/utils.py:70: UserWarning: "ImageToTensor" pipeline is replaced by "DefaultFormatBundle" for batch inference. It is recommended to manually replace it in the test data pipeline in your config file.
  'data pipeline in your config file.', UserWarning)


array([[4.0032205e+03, 3.9050674e+03, 4.4722310e+03, 4.4629087e+03,
        5.7059830e-01],
       [4.7793750e+03, 4.9904180e+03, 5.0820996e+03, 5.6726792e+03,
        5.6033736e-01]], dtype=float32)

*detections* is a list of all the detected worm sections. Every detection is a 5-tuple of the form *(xmin, ymin, xmax, ymax, score)*, i.e. the location of the bounding box followed by the confidence.

**Exercise:**
- Plot the predicted worm boxes. You may use the code provided in tutorial 2.
- Load the model weights of a different checkpoint, e.g. after the first epoch. What are the differences?

# Evaluating the model

To get an idea how well we're doing we can run the same evaluation function that will be used for the leaderboard on our example file.
For this we first need to convert to model output into the format that is readable by our scoring function.

In [19]:
columns = ['section_id', 'file_name', 'xmin', 'xmax', 'ymin', 'ymax', 'detection_score']
df_results = pd.DataFrame(columns=columns)

predictions = []

# Loop over all detected boxes and convert the information into a dictionary
for box in detections[0]:
    xmin, ymin, xmax, ymax, score = box
    xmin, ymin, xmax, ymax = int(xmin), int(ymin), int(xmax), int(ymax)  # Convert predicted coordinates to integer values
    box_dict = dict(
        section_id=f'{example_image}@{xmin}-{xmax}-{ymin}-{ymax}',
        file_name=example_image,
        xmin=xmin,
        ymin=ymin,
        xmax=xmax,
        ymax=ymax,
        detection_score=score
    )
    predictions.append(box_dict)

# Convert the dictionary into a dataframe with section_id as index
prediction_df = pd.DataFrame(predictions)
prediction_df.set_index('section_id', inplace=True)
prediction_df.head()

file_name  xmin  ymin  xmax  ymax  detection_score
section_id                                                                    
1_D.jpg@4003-4472-3905-4462   1_D.jpg  4003  3905  4472  4462         0.570598
1_D.jpg@4779-5082-4990-5672   1_D.jpg  4779  4990  5082  5672         0.560337
1_D.jpg@2930-3302-7185-7667   1_D.jpg  2930  7185  3302  7667         0.536524
1_D.jpg@5344-5884-2749-3272   1_D.jpg  5344  2749  5884  3272         0.533923
1_D.jpg@4516-4842-1575-1984   1_D.jpg  4516  1575  4842  1984         0.529677

Next, we restrict our dataset to only the file for which we made predictions (*1_D.jpg*) and run the score function

In [20]:
ground_truth = section_df.loc[section_df.file_name==example_image]

In [21]:
evaluator = PredictionEvaluator(ground_truth)
thresholds = [0.5, 0.6, 0.7]

For the leaderboard, the predicted worm section boxes are compared to the actual worm section boxes. We compute the [Intersection over Union (IOU)](https://towardsdatascience.com/iou-a-better-detection-evaluation-metric-45a511185be1) between the boxes. If the value is greater than a threshold (e.g. 0.5) the prediction counts as a match.

In [22]:
get_leaderboard_score(prediction_df, thresholds, evaluator)

100%|██████████| 1/1 [00:00<00:00, 25.18it/s]


{'detection_acc@iou0.5': 27.27,
 'detection_tp@iou0.5': 30,
 'detection_fp@iou0.5': 70,
 'detection_fn@iou0.5': 10,
 'detection_acc@iou0.6': 26.13,
 'detection_tp@iou0.6': 29,
 'detection_fp@iou0.6': 71,
 'detection_fn@iou0.6': 11,
 'detection_acc@iou0.7': 16.67,
 'detection_tp@iou0.7': 20,
 'detection_fp@iou0.7': 80,
 'detection_fn@iou0.7': 20,
 'score': 70.07}

The function *get_leaderboard_score* gives us a lot of information that we can use to debug our model. 
- *score* is the score that will be shown in the leaderboard. It is the sum of the individual detection_acc@X values.
- Additionally, for each IOU threshold the *accuracy, true positives, false positives* and *false negatives* are computed.

**Exercise:**
- Which type of error is the most common? How could you combat this?
- The output shows true positives (tp), false positives (fp), false negatives (fn) but no true negatives. Why is that?
- Evaluate the model on the *dummy_test* dataset we created previously.

# A first submission

Now, we have all the pieces in place to create our first submission. We will predict the worm sections for all files and store them in a DataFrame.
We will need to provide a prediction for all the files that are listed in *test_files.csv*. Let's load the file first.

In [23]:
files = pd.read_csv(f'{data_folder}/test_files.csv', sep=';', header=None)
file_names = files[0].values
file_names

array(['100_D.jpg', '100_C.jpg', '100_B.jpg', '100_AA.jpg', '100_A.jpg',
       '101_DD.jpg', '101_C.jpg', '101_B.jpg', '101_AA.jpg', '101_A.jpg',
       '86_D.jpg', '86_C.jpg', '86_B.jpg', '86_AA.jpg', '86_A.jpg',
       '88_D.jpg', '88_C.jpg', '88_B.jpg', '88_A.jpg', '89_D.jpg',
       '89_C.jpg', '89_B.jpg', '89_AA.jpg', '89_A.jpg', '90_D.jpg',
       '90_C.jpg', '90_B.jpg', '90_AA.jpg', '90_A.jpg', '91_D.jpg',
       '91_C.jpg', '91_B.jpg', '91_AA.jpg', '91_A.jpg', '92_D.jpg',
       '92_C.jpg', '92_B.jpg', '92_AA.jpg', '92_A.jpg', '93_D.jpg',
       '93_C.jpg', '93_B.jpg', '93_AA.jpg', '93_A.jpg', '94_D.jpg',
       '94_C.jpg', '94_B.jpg', '94_AA.jpg', '94_A.jpg', '95_D.jpg',
       '95_C.jpg', '95_B.jpg', '95_AA.jpg', '95_A.jpg', '96_D.jpg',
       '96_C.jpg', '96_B.jpg', '96_AA.jpg', '96_A.jpg', '97_D.jpg',
       '97_C.jpg', '97_B.jpg', '97_A.jpg', '98_D.jpg', '98_C.jpg',
       '98_B.jpg', '98_AA.jpg', '98_A.jpg', '99_D.jpg', '99_C.jpg',
       '99_B.jpg', '99_AA.jpg', '99_A.j

For simplicity, we created a function that runs the above code on all file names and returns a dataframe with the predictions. This will take around five minutes.

In [24]:
prediction_df = create_predictions(file_names, cfg, checkpoint, device='cuda')

load checkpoint from local path: /root/data/AmazonSageMaker-gdsc-tutorials/data/tutorial_exps/epoch_3.pth


100%|██████████| 73/73 [01:22<00:00,  1.14s/it]


In [25]:
prediction_df.head()

file_name  xmin  ymin  xmax  ymax  \
section_id                                                         
100_D.jpg@664-851-3168-3777    100_D.jpg   664  3168   851  3777   
100_D.jpg@707-904-919-1386     100_D.jpg   707   919   904  1386   
100_D.jpg@1259-1454-4218-4594  100_D.jpg  1259  4218  1454  4594   
100_D.jpg@1177-1748-2206-2716  100_D.jpg  1177  2206  1748  2716   
100_D.jpg@2788-2964-6293-6569  100_D.jpg  2788  6293  2964  6569   

                               detection_score  
section_id                                      
100_D.jpg@664-851-3168-3777           0.540133  
100_D.jpg@707-904-919-1386            0.532225  
100_D.jpg@1259-1454-4218-4594         0.484710  
100_D.jpg@1177-1748-2206-2716         0.472370  
100_D.jpg@2788-2964-6293-6569         0.471982

Looks good! All that's left is to save the results to a csv and upload it on the [GDSC website](https://gdsc.ce.capgemini.com/).

In [26]:
prediction_df.to_csv(f'{data_folder}/results_tutorial3.csv', sep=';')

This submission should lead to a score of around 28. 

**Exercise:**
- The submission score is a lot less than the score of 70.07 we computed above. What are potential reasons for this?

# Summary

This notebook covered a LOT of different topics. We covered how to 
- Train a basic object detection model
- Use a trained model to create predictions on a nodule image
- How to evaluate the predictions
- How to create a first submission

all done on a dummy dataset. With this, we have the technical foundations to creating a good submission!

In the next tutorial, you will learn how to run and evaluate the model from this notebook on the complete dataset as a Sagemaker training job.

**REMINDER: Remember to shut down the *ml.g4dn.xlarge* instance when you aren't using it.**